src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1432,41 +1464,175 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
 int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp,
                                                    optional_yield y, uint32_t flags)
 {
-  /*
-  1. Check if object exists in Object Directory
-  2. get dirty flag and version to construct the oid in cache correctly 
-  3. loop and delete all blocks in the cache
-  4. delete the head block
-  5. Update the in memory data structure for all blocks, and update the block directory also
-  6. Update the in memory data structure for head block, and update the block directory also
-  7. If the blocks reside in other caches, send remote request for the same
-  8. Need to figure out a way to get all versions to be deleted in case of versioned objects when a version is not specified.
-  9. If the object is not in object directory call next->delete_obj
-  */
-  rgw::d4n::CacheObj obj = rgw::d4n::CacheObj{ // TODO: Add logic to ObjectDirectory del method to also delete all blocks belonging to that object
-			     .objName = source->get_key().get_oid(),
-			     .bucketName = source->get_bucket()->get_name()
-			   };
+  // TODO: 
+  // 1. Send delete request to cache nodes with remote copies
+  // 2. See if we can derive dirty flag from the head block 
+  // 3. Add lock so cleaning method doesn't remove "D_" prefix
 
-  int ret = -1;
-  if ((ret = source->driver->get_obj_dir()->del(dpp, &obj, y)) < 0) 
-    ldpp_dout(dpp, 0) << "D4NFilterObject::" << __func__ << "(): ObjectDirectory del() method failed, ret=" << ret << dendl;
+  rgw::sal::Attrs attrs;
+  std::string head_oid_in_cache;
+  rgw::d4n::CacheBlock block;
+  if (!source->check_head_exists_in_cache_get_oid(dpp, head_oid_in_cache, attrs, block, y) && !block.deleteMarker) {
+    ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterDeleteOp::" << __func__ << "(): calling next delete_obj" << dendl;
+    return next->delete_obj(dpp, y, flags);
+  } else {
+    int ret = -1;
+    bool objDirty = false;
+    auto blockDir = source->driver->get_block_dir(); 
+    std::string version, policy_prefix;
+
+    if (!source->get_bucket()->versioned()) {
+      version = source->get_object_version(); 
+    } else if (source->get_bucket()->versioned() && !source->have_instance()) {
+      rgw::d4n::CacheBlock deleteBlock;
+      block.prevVersion = std::pair<std::string, bool>(block.version, block.deleteMarker);
+      block.deleteMarker = true;
+
+      if (source->get_bucket()->versioned() && !source->get_bucket()->versioning_enabled()) { // if versioning is suspended
+        block.version = "null"; 
+      } else {
+	// create a delete marker
+	enum { OBJ_INSTANCE_LEN = 32 };
+	char buf[OBJ_INSTANCE_LEN + 1];
+	gen_rand_alphanumeric_no_underscore(dpp->get_cct(), buf, OBJ_INSTANCE_LEN);
+        block.version = buf; // using gen_rand_alphanumeric_no_underscore for the time being
+      } 
+      
+      deleteBlock = block;
+      deleteBlock.cacheObj.objName = "_:" + deleteBlock.version + "_" + deleteBlock.cacheObj.objName;
```

Comment:
can't source->get_oid() be used here?

Reply:
Not in this case, because this is for the situation where there is no version provided to the request, which means the oid would not contain the version.